qiskit/circuit/register.py

```diff
@@ -149,18 +185,33 @@ def __iter__(self):
 
     def __eq__(self, other):
         """Two Registers are the same if they are of the same type
-        (i.e. quantum/classical), and have the same name and size.
+        (i.e. quantum/classical), and have the same name and size. Additionally,
+        if either Register contains new-style bits, the bits in both registers
+        will be checked for pairwise equality. If two registers are equal,
+        they will have behave identically when specified as circuit args.
 
         Args:
             other (Register): other Register
 
         Returns:
             bool: `self` and `other` are equal.
         """
-        try:
-            return self._repr == other._repr
-        except AttributeError:
-            return False
+        if self is other:
+            return True
+
+        res = False
+        if type(self) is type(other) and \
+                self._repr == other._repr and \
+                all(
+                        # For new-style bits, check bitwise equality.
+                        sbit == obit
+                        for sbit, obit in zip(self, other)
+                        if None in
+                        (sbit._register, sbit._index,
+                         obit._register, obit._index)
+                ):
```

Comment:
I know I asked this in an earlier review, but was there performance impact here. In #5272 we removed the bit wise equality check because it had significant overhead for DAGNode creation. I don't think it will come up because it was per bit iirc and only an artifact of `Bit.__eq__` previously comparing registers, but it's just something we should keep in mind.

Reply:
Agree, much of the transpiler has been updated to avoid register comparisons in #5515 so it's unlikely to be an issue, but agree its something to keep an eye on.